Add type refinement for Union

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* #57013
* #57012
* __->__ #57011
* #53180

Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

Commit:
Add type refinement for Union

[ghstack-poisoned]

eellison commented on torch/csrc/jit/frontend/convert_to_ssa.cpp:
```diff
@@ -93,10 +93,7 @@ struct ControlFlowLoadStores {
-      if (!unified) {
-        continue;
-      }
+      auto unified = unifyTypes(true_type, false_type, true);
```
nit: `/*default_to_union*/`

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1167,14 +1167,24 @@ struct to_ir {
     }
+    if (const auto union_type = lhs_value->type()->cast<UnionType>()) {
+      std::vector<TypePtr> present_types =
+          union_type->set_difference_of({NoneType::get()});
```
nice!

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1167,14 +1167,24 @@ struct to_ir {
+    if (const auto union_type = lhs_value->type()->cast<UnionType>()) {
+      std::vector<TypePtr> present_types =
+          union_type->set_difference_of({NoneType::get()});
+      Refinement present(name, UnionType::create(present_types));
```
What happens when you pass in pass in a vector of size one here ? Does it create a single element Union (which is not good). This is sort of a general question of, what should the return type of `UnionType::create` be ? Should we enforce that the caller maintains the invariant that there are multiple unique types, or should we throw if there is a single type, or return a non UnionType ?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1648,7 +1658,19 @@ struct to_ir {
+      SugaredValuePtr true_val = save_true->findInParentFrame(x);
+      SugaredValuePtr false_val = save_false->findInParentFrame(x);
+      bool default_to_union = false;
+      if (true_val && false_val) {
```
I'm confused by this logic. We're checking if both values were already defined in the parent scope and if they both are then if both types are equal to each other we allow them unify to union ? If both types are equal to each other, shouldn't they already unify ? 

I'm not sure how whether or not they were defined in the parent scope should matter. Plus, `true_val` should always equal `false_val` since both save_true and save_false have the same parent scopes.

ansley replied:
Good catch. This was me misunderstanding the API, but I have it fixed now. What I really wanted was to check the type in `environment_stack->type_table_` if it exists. It could be the case that a variable is annotated as being a Union but is refined in the if-else branches, e.g. 

```python
def fn(b: bool):
    y: Union[int, str, None] = None
    if (b):
        # `y` should be a `str` in this branch without needing an
        # `isinstance` check
        y = "foo"
    else:
       # same here
        y = 1
    # But outside of the if-else branche, whatever the value 
    # is (whether `"foo"` or `1`) should go back to being a Union
    # and would need an `isinstance` for type refinement
```

ansley replied:
Thanks, I was using the wrong function in our API. It should be fixed now, but please take a look at it again just to make sure.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1751,16 +1773,30 @@ struct to_ir {
+        auto union_type = val->type()->expect<UnionType>();
+        std::vector<TypePtr> not_isinstance_types =
+            union_type->set_difference_of({gathered.types.at(0)});
+        if (not_isinstance_types.size() == 1) {
```
This brings up the question above of api

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1751,16 +1773,30 @@ struct to_ir {
+
+      if (val->type()->kind() == UnionType::Kind) {
+        auto union_type = val->type()->expect<UnionType>();
+        std::vector<TypePtr> not_isinstance_types =
```
Isn't the code here when `isinstance` has worked successfully ? Don't we want all types in the Union that are subtypes of `isinstance`?

ansley replied:
The RefinementSet needs to have "true refinements" and "false refinements". I'm creating the "false refinements" here.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1751,16 +1773,30 @@ struct to_ir {
+      if (val->type()->kind() == UnionType::Kind) {
+        auto union_type = val->type()->expect<UnionType>();
+        std::vector<TypePtr> not_isinstance_types =
+            union_type->set_difference_of({gathered.types.at(0)});
```
`set_difference_of` implies we are just subtracting types. I think in the use cases of the API in this file, what we want is to either take all types that are subtypes of a particular type, or take all types that are not subtypes of a particular type.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3303,6 +3341,14 @@ struct to_ir {
+        // correctly sets the value's type to match the variable's
+        // Union annotation
+        if (type_hint && type_hint->kind() == UnionType::Kind) {
+          environment_stack->setType(Var(tree).name().name(), type_hint);
```
this api is bypassing the safety checks. I think we should be using setSugaredVar or something along those lines. I'm also a little confused about what we're trying to achieve here. Do you mind commenting on the use case ?

eellison reviewed (commented):
Looks good, thanks for breaking it up! I have a few questions.

Commit:
Update on "Add type refinement for Union"




[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

eellison commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -24,6 +24,17 @@ struct Function;
+// the serialized representation (slower, but guaranteed to break ties).
+// This is a struct so that it can be used in certain std container
+// types.
+struct TypeEqual {
```
this isn't correct for Unions/Optionals. In general we should be wary of code that attempts to optimize for negligible perf gains but adds complexity and maintenance to the system

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -90,7 +91,7 @@ bool isMutableTypeImpl(
   auto kind = type->kind();
   if (kind == TypeKind::TensorType || kind == TypeKind::ListType ||
-      kind == TypeKind::ClassType || kind == TypeKind::DictType) {
+      kind == TypeKind::ClassType || kind == TypeKind::DictType || kind == TypeKind::UnionType) {
```
We shouldn't have Union here for the same reason we don't have Optional here. If it is a union of immutable types than the type itself is also immutable.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3276,7 +3369,9 @@ struct to_ir {
     // x : Any = 1 if cond else "str"
-    if (type_hint == AnyType::get() && out_val->type() != AnyType::get()) {
+    if (type_hint &&
+          ((type_hint == AnyType::get() && out_val->type() != AnyType::get()) ||
```
this was valid because you can always cast something to Any, it's not valid to always cast something to Union

ansley replied:
Changing this to "If the type hint is Union and the output is a subtype of that Union, then insert an unchecked cast". Let me know if that works

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1165,14 +1188,23 @@ struct to_ir {
       }
     }
+    if (const auto union_type = lhs_value->type()->cast<UnionType>()) {
+      UnionTypePtr remaining = union_type->withoutNone();
```
Do we have any other callers of `withoutNone` ? This seems like a very specific API to add to Union Types if this is the only caller. I would expect this to just be something like `union_type->subtractTypeSet(None)`

ansley replied:
Fixing in previous PR

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1652,6 +1683,16 @@ struct to_ir {
 
+      // If the variable we're looking at is known to be Union[T1, T2],
+      // then it's okay to have one branch return T1 and the other
+      // return T2. (We're not actually going to use `unified` again;
```
`We're not actually going to use `unified` again;` seems misleading, it will be used again.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1652,6 +1683,16 @@ struct to_ir {
+      // it's acting as a flag)
+      c10::optional<TypePtr> full_true_type = environment_stack->getType(x);
+      c10::optional<TypePtr> full_false_type = environment_stack->getType(x);
+      if (full_true_type && full_false_type) {
```
Can you explain why `unifyTypes(tv->type(), fv->type());` call above does not handle whatever case this has been added to cover ?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
-        gathered.types.at(0)->isSubtypeOf(val->type()) &&
-        obj.kind() == TK_VAR) {
+
+    TORCH_CHECK(!gathered.types.empty(), "`isinstance` must be used with a "
```
this is technically valid syntax... `isinstance(3, ())` it just always returns False

eellison reviewed (commented):
Looking good, I still need to review the changes in `emitIsInstance`

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
+                                        gathered.types[0] :
+                                        UnionType::create(gathered.types);
+
+      Refinement isinstance(ident, single_comparison);
```
is this used at all ?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
+          }
+          // Check all types in `gathered.types` to see if there's a
+          // common supertype
+          for (const TypePtr comparison_type : gathered.types) {
```
im a little confused by this logic, why are we unifying the comparison_type with the type?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
+          // common supertype
+          for (const TypePtr comparison_type : gathered.types) {
+            c10::optional<TypePtr> unified = unifyTypes(type, comparison_type);
+            if (!unified) {
```
Would you mind commenting with what the algorithm we're trying to do here is ? I'm having trouble following the logic.

ansley replied:
I'm guessing that the confusing part is only if the lhs is a `Union`, but lmk if that's not the case

For example:
```
def fn(x: Union[int, str]):
    if isinstance(x, str):
        call_str_method(x)    # x should be refined to `str` and this method should work
    …
```

This is the general algorithm:
1. Get single comparison type from rhs. (Note: we only ever have one type, but, in the next few steps, I'm going to mostly use "type" to refer to what we get from a single-element comparison, and "types" to refer to a 2+-element comparison. Of course, the latter becomes a single `Union` type.)
2. Put that comparison type in the `isinstance` vector, which says “if the lhs type is an instance of one of the rhs types, then we know we can refine it to one of the rhs types”. (We can’t know which exact rhs type the lhs type can be refined to until runtime; we can only generate code saying that it must be one of those types. A cool follow-up project would be automatically generating hidden conditional asserts to further refine the type.)
3. For each type in the lhs `Union`, see if that type does NOT share a common parent with the rhs comparison type, i.e. returns `c10::nullopt` from `unifyTypes`. This allows us to refine Unions to smaller Unions.
4. Create the correct Refinements.

eellison replied:
> 1. Get single comparison type from rhs. 

I'm not sure this is helpful. As in the following case:
```
def foo(x: Union[int, str]):
     if isinstance(x, (int, float)):
```
`Union[int, float]` doesn't subtype `Union[int, str]` , however we would like to retain the `isinstance` check and insert a runtime check that the input is an int with a corresponding refinement to int. 

I think it might be easier to reason about this if we decompose Union inputs into their composed types. This also unifies the type logic with the other path. 

for each type in the input, we want to see if it subtypes any of the types in `isisntance`:

* for the types that do subtype types in `isinstance`, we record which types are their supertype, and add those supertypes to the runtime checks we need to generate in the `isinstance`. the union of all the types that are subtypes will be the type refinement in the true branch.
* The union of types that do not subtype types in `isinstance` get added to the false branch type refinements

If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.


> A cool follow-up project would be automatically generating hidden conditional asserts to further refine the type 
I'm not sure what you mean here. `prim::UncheckedCast` is what refines the type in the IR if that's what you mean ?

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
+                                        [&](const TypePtr t) {
+                                          return t->isSubtypeOf(val->type());
+                                        });
+    if (all_are_subtypes && obj.kind() == TK_VAR) {
```
Could you explain the `# x must be int here` line?

eellison replied:
we only trigger the union logic if every type in `isinstance` is a subtype of `x`, which in this case would be False, so we won't generate any reginements here, even though we would like to refine `Union[int, str]` to `int` here.

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
+      // logic for single vs. multiple comparisons
+      const TypePtr single_comparison = gathered.types.size() == 1 ?
+                                        gathered.types[0] :
+                                        UnionType::create(gathered.types);
```
Good point about using `unifyTypeList`

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
           }
+        };
+
+    auto dedup = [&](std::vector<TypePtr>& types) {
```
This clang-tidy warning seems serious. Could you use this `dedup` in code and also add test case to make sure if `dedup` is not called that test fails.

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
-          if (actual_type->isSubtypeOf(typ)) {
-            return true;
+
+          if (t->kind() == UnionType::Kind) {
```
For future clean up:
Consider maybe adding another public common API on base type so that we don't need to gut out internal types of `union`, `optional` and `number` like this. Instead we can call that new API to perform type computations (like can A be refined into B).

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+            if (lhs_type->isSubtypeOf(rhs_type)) {
+              return false;
+            }
+            if (rhs_type->isSubtypeOf(AnyListType::get()) &&
```
Future cleanup:

`AnyListType` should be a super type of all `ListType`

gmagogsfm commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -48,6 +48,7 @@ class MutableTypePtrHelper {
       case TypeKind::DictType:
       case TypeKind::ClassType:
       case TypeKind::TensorType:
+      case TypeKind::UnionType:
```
@eellison What's the possible consequence of landing this line? Is it a correctness issue or just perf regression? I learned from Ansley that this is temporary for this PR only and will be fixed by later PRs in the stack.

eellison replied:
correctness issue

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
+      auto unified = unifyTypes(
+          tv->type(), fv->type(), /*default_to_union=*/default_to_union);
+
+      // Special logic for if we have a Union-typed value: check if
```
This does not follow with the rest of our type checking. 
```
def foo(x: Optional[int]) -> int:
    if x is None:
         x = 4
    return x
```
It is important to be able to refine types. We should remove this logic.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
+      // `unified` to be the known type to prevent unintentional
+      // narrowing of the Union type. E.g. `Union[T1, T2, T3]` should
+      // not become `Union[T1, T2]` just because its initial definition
+      // is as "`T1` if cond else `T2`". If the Union we got from type
```
The Union is guaranteed to be a subtype of its original type, since we only allow redeclaration/redefinition of variables if they subtype the existing variable in the parent frame.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
-      auto unified = unifyTypes(tv->type(), fv->type());
+      // If we already know that the type is a Union, then we want to
+      // allow heterogeneous types to be unified
+      c10::optional<TypePtr> full_type = environment_stack->getType(x);
```
So, the intent of this logic is to allow:
```
def foo(x: Union[int, str]):
    if cond:
        x = 2
    else:
        x = "str"
   return x
```
but not allow
 ```
def foo():
    if cond:
        x = 2
    else:
        x = "str"
   return x
```

Could you explain the intention behind this?

ansley replied:
Yes, that's exactly what I'm trying to do. If we allowed the latter example, then I think we'd fail to catch some legitimate errors. I like the idea of only allowing `Union` if the user explicitly annotates it

cc @gmagogsfm

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
-          for (Expr e : TupleLiteral(classinfo).inputs()) {
-            gather(e);
+    // For use with `unifyTypeList`
+    std::stringstream nowhere;
```
can you declare this closer to where it is used ? it's a little confusing to see it here. or, make unifyTypeList take in an optional `why_not` parameter

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+    // expression, e.g. `isinstance(x, (float, int))` -> `{float, int}.
+    // While doing this, we also want to decompose any `Union` types,
+    // namely `Union`, `Optional`, and `Number`
+    // NB: We can't use `auto` with this lambda because it contains a
```
nit: Typically, NB is something that whoever reading the code needs to make sure to think about. IMO whether the lambda is defined with auto or a manual declaration isn't that necessary for the reader to think about

ansley replied:
Gotcha, I had assumed it stood for [nota bene](https://en.wikipedia.org/wiki/Nota_bene) and had been using it like that. Thanks for correcting me :)

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+    std::function<void(const Expr&, std::vector<TypePtr>&, bool)> gather_types;
+
+    gather_types =
+        [&](const Expr& expr, std::vector<TypePtr>& types, bool is_lhs) {
```
note that you can use `auto` with `gather_types =
        [&](const Expr& expr, std::vector<TypePtr>& types, bool is_lhs) -> void` i'm pretty sure

ansley replied:
You can't. Please see note above. Example compiler error:

```
../torch/csrc/jit/frontend/ir_emitter.cpp: In lambda function:
../torch/csrc/jit/frontend/ir_emitter.cpp:1756:17: error: use of ‘gather_types’ before deduction of ‘auto’
                 gather_types(e, types, is_lhs);
                 ^~~~~~~~~~~~
```

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+    // recursive call
+    std::function<void(const Expr&, std::vector<TypePtr>&, bool)> gather_types;
+
+    gather_types =
```
The only shared logic here between lhs and rhs is the decomposition of Union types. I think this would be easier to follow if you factored that part out into its own function and then separately wrote out logic for lhs/rhs.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+
+    Value* lhs_val;
+
+    // Get the types on the right-hand side of the `isinstance`
```
What do you think about writing at the top of this function a description of the algorithm ? 

> we also want to decompose any `Union` types
this describes what we are doing, but it doesn't describe the motivation

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+      // false branch, not `(List[str], str, int)` in the true branch
+      // and `None` in the false branch
+      for (TypePtr lhs_type : lhs_types) {
+        if (lhs_type == AnyType::get()) {
```
I think we might be able to rewrite this to not special case Any if we iterated through the types on the rhs instead of the lhs, but i have to think about it more..

eellison replied:
This is evident that we're doing the logic wrong here. I think we want to check if lhs is a subtype of rhs or rhs is a subtype of lhs. However, we only insert refinements for the types in the rhs that subtype lhs.

ansley replied:
I think the only line that should be changed is `return lhs_type->isSubtypeOf(rhs_type)`, which should be `return rhs_type->isSubtypeOf(lhs_type)`. This definitely was a mistake, so thanks for catching it.

However, I don't understand why we'd want to check that `lhs` is a subtype of `rhs` or `rhs` is a subtype of `lhs`--wouldn't we introduce correctness issues if we allowed for "widening" refinements?

For example, consider:
```
def fn(x: Optional[int]):
    if isinstance(x, int):
        # "narrowing" refinement, aka the rhs is a subtype of the lhs
        # We know `x` is an int here and can use it as an int
```
in comparison to:
```
def fn(x: int):
    if isinstance(x, Optional[int]):
        # "widening" refinement, aka the lhs is a subtype of the rhs
        # This is bad! We already know `x` is an int; we shouldn't
        # insert casts so that it can now hold `None` instead
```

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+      }
+
+      // Get a single type for the true and false branches
+      unified_true = isinstance_types.empty() ? nullptr
```
nit: 
`if (unified_true)` -> `if (isinstance_types.empty())` directly use predicate instead of introducing unnecessary temporary `nullptr`

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+                not_isinstance_types,
+                nowhere,
+                /*default_to_union=*/true);
+      if (unified_false) {
```
same here

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    if (obj.kind() == TK_VAR) {
```
we only do this logic for `TK_VAR`, but don't we also want to do the filtering of isinstance types even for unnamed expressions ?

ansley replied:
I might be misreading the original code, but it seems like that's what we do [there](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/ir_emitter.cpp#L1742), as well. The `RefinementSet` `refinement` will only have something in it if the lhs is a `Var`

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+      refinement = RefinementSet(true_refinements, false_refinements);
+    }
+
+    bool is_statically_true =
```
this seems like work that should have already been accomplished in our earlier pass. Do you mind explaining what didnt work  or was less preferable than current version about 
> for each type in the input, we want to see if it subtypes any of the types in isisntance:
for the types that do subtype types in isinstance, we record which types are their supertype, and add those supertypes to the runtime checks we need to generate in the isinstance. the union of all the types that are subtypes will be the type refinement in the true branch.
The union of types that do not subtype types in isinstance get added to the false branch type refinements
If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.

eellison replied:
> If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.

We're recreating work here that I think would already be done is all, but ya can leave for follow up if there is a cleanup that is possible.

eellison reviewed (commented):
Looks great !! I had some comments about cleanup of the refinement algorithm, but I think that can probably be deferred until follow up.

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -315,6 +315,13 @@ struct Environment {
   }
 
+  c10::optional<TypePtr> getType(const std::string& name) {
+    if (type_table.find(name) != type_table.end()) {
```
I noticed that the existing API inserts a `Load` if it finds the type, which wasn't what I wanted. I can't see how it wouldn't hurt anything, though--wouldn't the `Load` just be optimized away later? What do you think?

eellison replied:
yes, it would get removed later.

ansley replied:
My question was more about whether it would be good practice to add extraneous `Load`s if we knew they'd be removed later

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1806,84 +1834,148 @@ struct to_ir {
+          found_refinement = lhs_type;
+        } else if (*maybe_smaller_type == *rhs_type) {
+          // We want the narrowest possible type
+          found_refinement = get_smaller_type(found_refinement, rhs_type);
```
So, I think this probably mostly works because of our limited type hierarchies, especially when you `standardizeVectorForUnion` before the call. 

However, i think the logic could be adjusted a bit. If you have multiple different rhs type which subtype a lhs type, the different rhs types do not necessarily have a type relation to each other. 
          t1
      /    |    \
t2       t3      t4

t2 <: t1
t3 <: t1
t4 <: t1

If we are doing isinstance(t1, (t2, t4)), when we process `t4` the refinement should be Union[t2, t4] instead of just t4. Right now `get_smaller_type(found_refinement, rhs_type)` will produce nullptr. 

I think we want to either initialize found_refinement if it hasn't be set yet, or set found_refinement to the union of  the existing type and rhs_type here. 
```
found_refinement = found_refinement ? unifyTypes(found_refinement, rhs_type) : rhs_type;
```

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1806,84 +1834,148 @@ struct to_ir {
+        // If the lhs couldn't be a subtype of the rhs (or couldn't
+        // be "refined" to itself, as in the `List[str]` and `list`
+        // case above), then we add `lhs_type` to the false branch
+        // refinements. This is because the type can still be itself
```
nice!!

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1189,11 +1190,16 @@ c10::optional<TypePtr> UnionType::subtractTypeSet(std::vector<TypePtr>& to_subtr
+  bool is_numbertype = false;
+  if (auto as_union = contained->cast<UnionType>()) {
+    is_numbertype = as_union->containedTypes().size() == 3 &&
+                    as_union->canHoldType(NumberType::get());
```
eventually it would be nice to replace `canHoldType` with `subtypeOf` and remove `canHoldType`, since it's a duplicate API of the more general one

eellison reviewed (approved):
looks great 😍 😍 😍 

Just one small comment about the type unification of multiple types on `rhs` that `lhs` subtypes which I think we should fix.

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572811).

Commit:
Update on "Add type refinement for Union"


Differential Revision: [D28572811](https://our.internmc.facebook.com/intern/diff/D28572811)

[ghstack-poisoned]
